### ISSUE TITLE
Tracking Issue for future-incompatibility lint `dollar_crate_in_matcher`

### ISSUE THREAD

**Issue by nik-rev** [rust-lang/rust#155123]
This is the **tracking issue** for the `dollar_crate_in_matcher` future-compatibility warning and other related errors. The goal of this page is to describe why this change was made and how you can fix code that is affected by it. It also provides a place to ask questions or register a complaint if you feel the change should not be made. For more information on the policy around future-compatibility warnings, see our [breaking change policy guidelines][guidelines].

[guidelines]: https://rustc-dev-guide.rust-lang.org/bug-fix-procedure.html

### What is the warning for?

Usage of `$crate` in matcher.

### Why was this change made?

`$crate` is inconsistent with the behavior of other keywords in matchers, namely that other keywords work like any other identifier, and are currently not reserved in this position.

### Example

```rust
#![deny(dollar_crate_in_matcher)]

macro_rules! m {
    ($crate) => {};
}
```

### Recommendations

Do not use `$crate` in matcher.

### When will this warning become a hard error?

Unknown

### Steps

- [ ] Implement the lint
- [ ] Raise lint level to deny
- [ ] Change the lint to report in dependencies
- [ ] Switch to a hard error

### Implementation history

- https://github.com/rust-lang/rust/pull/155121
